_posts/en/2016-xx-xx-Implementing-asynchronous-loops.md

```diff
@@ -0,0 +1,77 @@
+---
+layout: post
+
+title: Implementing asynchronous loop
+tip-number: xx
+tip-username: madmantalking
+tip-username-profile: https://github.com/madmantalking
+tip-tldr: You may run into problems while implementing asynchronous loops. 
+
+categories:
+    - en
+---
+
+Let's try out writing an asynchronous function which prints the value of the loop index every second.
+
+```js
+for (var i=0; i<5; i++) {
+	setTimeout(function(){
+		console.log(i); 
+	}, 1000);
+}  
+```
+The output of the above programs turns out to be
+```js
+> 5
+> 5
+> 5
+> 5
+> 5
+```
+So this definitely doesn't work.
+
+**Reason**
+
+Each timeout refers to the original `i`, not a copy. So the for loop increments `i` until it gets to 5, then the timeouts run and use the current value of `i` (which is 5).
+
+Well , this problem seems easy. An immediate solution that strikes is to cache the loop index in a temporary variable.
+
+```js
+for (var i=0; i<5; i++) {
+	var temp = i;
+ 	setTimeout(function(){
+		console.log(i); 
+	}, 1000);
+}  
+```
+But again the output of the above programs turns out to be
+```js
+> 5
+> 5
+> 5
+> 5
+> 5
+```
+So , that doesn't work either , because blocks don't create a scope and variables initializers are hoisted to the top of the scope. In fact, the previous block is the same as:
+```js
+var temp;
+for (var i=0; i<5; i++) {
+ 	temp = i;
+	setTimeout(function(){
+		console.log(i); 
+  	}, 1000);
+}  
+```
+**Solution**
+
+There are a few different ways to copy `i`. The most common way is creating a closure by declaring a function and passing `i` as an argument. Here we do this as a self-calling function.
+```js
+for (var i=0; i<5; i++) {
+	(function(num){
+		setTimeout(function(){
+			console.log(num); 
+		}, 1000); 
+	})(i);  
+}  
```

Comment:
@gogainda 
"Function declarations should not be placed in blocks. Use a function expression or move the statement to the top of the outer function" - jslint